draft.md

```diff
@@ -0,0 +1,99 @@
+# Draft IPLD Unixfs Spec
+
+## Basic Structure
+
+A Unixfs is either a file or a directory.
+The top level IPLD object is a CBOR map with at least two fields: `type` and `data`
+and maybe a few other such as a version string or a set of flags.
+The `type` field is either `file` or `dir`.
+
+## IPLD `file`
+
+If an IPLD file is a leaf its CID type is `raw` (0x55) and has no structure.
+Otherwise its CID type is `dag-cbor` (0x71).
+The `type` field is set to `file` and the `data` field is an CBOR array.
+Each element of the array is CBOR map with the following fields:
+
+  - `data`: link
+  - `size`: cumulative size of `data`
+  - `fsize`: (file size) cumulative size of the payload of `data`
+  
+The `fsize` field is omitted if the link is `raw` as it is the same value as size.
+
+## IPLD `dir`
+
+An IPLD `dir` represents a directory.
+Its CID type is `dag-cbor` (0x71).
+The `type` field set to `dir` and the data field is an CBOR map.
+The key of the map is a filename and is a CBOR text string encoded in UTF-8.
+The value of the map is another CBOR map with the following standard fields:
+
+  - `type`
+  - `exe`: CBOR boolean: executable bit
```

Comment:
I think instead of just having executable here, we should do a full rwxrwxrwx unix permissions set (a uint32)

Reply:
The problem is the full unix set of permissions does not have a lot of meaning on other operating systems.  Even within unix systems it has limited meaning when stored in an archive.

Others may have stronger opinions on this than me.

Reply:
> I think instead of just having executable here, we should do a full rwxrwxrwx unix permissions set (a uint32)

@whyrusleeping the full `st_mode` ( entry type + permissions ) is in fact only 16 bits ( within a typically-32-bit-aligned struct member ). If we reuse it as-is we gain some extra bit of interop with everything that understands `st_mode` ( `git` took this path: https://stackoverflow.com/questions/737673/how-to-read-the-mode-field-of-git-ls-trees-output/8347325#8347325 ).

Of course this makes direct-query of type a bit harder, but then again every libc provides `S_IF...` macros

Reply:
I think there is no harm of having this additional data stored. Systems that have no meaning of those bits will skip them, systems that have will use them and allow for preservation.

It is very similar with uid and gid. They have no meaning on some systems, they may have no meaning on different machine with same system (different uid/gid mappings) but they are crucial if I wanted to, for example in future, use IPFS for `/home` storage in managed multi-user multi-workstation environment.

Reply:
I think that we should support the full range, but maybe not change the default behaviour?
or maybe we only record user executable by default.

thinking about it a bit more, the 'readable' flag really doesnt make a lot of sense in this context. I can read anything thats in ipfs...

Reply:
Storing the full `st_mode` by default just feels wrong to me and I will go as far as it could create additional complications down the road (what those complications are I am not sure).

One possible complication is how to handle the writable bit in `st_mode`, should it always be set or not by default.  If it is not set then should the st_mode be honored when extracting files.  If it is then that could be annoying as all files will end up readonly.  Or maybe only the executable bit should be honored by default, in that case it just seams better just to store that single bit.

For this version of the standard I feel rather strongly we should stick to just the executable bit as it was stated in the requirements (#1), or nothing at all (as @lgierth suggested we don't add additional meatadata).  The full `st_mode` can be included in a later version of the standard.